### ISSUE TITLE
Improve memory usage of histogram() aggregation

### ISSUE THREAD

**Issue by electrum** [prestodb/presto#8553]
There are two areas of improvement in `TypedHistogram`:

1. It uses a `LongBigArray` which uses ~20k per instance. This means each group takes up 20k even if there is only a single item in the histogram.
2. It stores a `BlockBuilder` which has a large overhead of single values. For example, `LongArrayBlockBuilder` uses 144 bytes.

**Comment by rash67** [comment]
@raghavsethi if you have thoughts on this. @electrum and I are looking into a state pattern that can optimize either the histogram impl and/or *BigArray impls to have special cases: they start out optimistic assuming a few elements (say <= 4) in very memory efficient ways. Once k is exceeded, we "upgrade" to the full version.

implementation details notwithstanding (an extra pointer, or maybe just a boolean), i should handle the degenerate case of a few distinct items in a group-by while not affecting larger sets

**Comment by rash67** [comment]
@nezihyigitbasi  @martint also mentioned you might be a goo poc.
there's sort of bifurcating options: tune the otheralll TypedHistory gram impl, or tart from scrach

ify you could provde  more context on why this way was chosen and which direction to take, that would help me immensely

I'm ON PTO today / monday, bur sporadic on github

thanks

**Comment by rash67** [comment]
tl;dr options
1. tweak Long//BigArray (or tweak to further control segment size and init segmetns (seems a little error-prone; though if w keep defaults, but use function heuristics to rigger lower-cost for small keysets and "upgrade" to full histogram when need be might win

2. roll our own long/big array. if we dont' mind paying a larger cost to grow
 we can use a high fill rate and reduce memory

3. pull back and rethink if w can redo a histograph in sow other fundamentally different way.
a, does it need to be exact?

4. LongBlockBuilder: not super familiar with this ,but a we again optimize space for small contants (C< 5? 10?) 

without knowing the input/output distributions, it's not clear what the biggest bang for buck is on each of these

@martint  intrested in your thought when you return

**Comment by raghavsethi** [comment]
Histogram needs to be exact.

Also, I think it's definitely worth exploring what the distribution of element count is while making this decision. You could either add some sort of summary stat (like people have done in the join operators), or try it out and see. If the idea is to try it out, then I would start with special casing the tiny states to see if there is improvement, and only considering a larger project if that does not yield good results.

**Comment by nezihyigitbasi** [comment]
I don't have much context on why we have the current implementation, it was added in #3062.

In general I agree with @raghavsethi that we need to collect some numbers before taking any decision and jumping to implementation. This will help us understand what's the biggest bang for the buck (I remember this was a user reported issue, it would be nice to get a heap dump and see what are the top consumers). Collecting numbers will also help us answer several other questions like: How common are small histograms? Are we optimizing for the common case? etc. etc.

**Comment by rash67** [comment]
I agree--in looking at this, I have no sense what impact on overall memory and cpu utilization any change I make would have.

I'll look into how we instrument and collect these #s and I'll have more questions

thanks

**Comment by Vayu** [comment]
Any progress on that?
We still have to use a seemingly less efficient workaround:
```sql
--HISTOGRAM(event) AS events_map,
-- workaround histogram runs out of memory
REDUCE(
    ARRAY_AGG(event),
    MAP(),
    (s, x) ->
        MAP_CONCAT(s, MAP(ARRAY[x], ARRAY[COALESCE(ELEMENT_AT(s, x), 0) + 1])),
    s -> s
) AS events_map,
```

**Comment by HunterVacui** [comment]
this still seems to be a problem as of today. Attempting to use HISTOGRAM on a column in a large table set causes my job to fail due to an out of memory issue, while using the REDUCE(ARRAY_AGG(... workaround posted above works
